aten/src/ATen/native/cuda/DepthwiseConv3d.cu

```diff
@@ -0,0 +1,655 @@
+}
+
+template <typename scalar_t,
+         int kKnownStrideH, int kKnownStrideW>
+__global__ void 
+conv_depthwise3d_cuda_backward_weight_kernel(
+    const PackedTensorAccessor32<scalar_t, 5> grad_output,
+    const PackedTensorAccessor32<scalar_t, 5> input,
+    PackedTensorAccessor32<scalar_t, 5> grad_kernel,
+    int dT, int dH_, int dW_,
+    int pT, int pH, int pW, 
+    int dilationT, int dilationH, int dilationW) {
+  const int kC = grad_kernel.size(0);
+  const int kT = grad_kernel.size(2);
+  const int kH = grad_kernel.size(3);
+  const int kW = grad_kernel.size(4);
+  
+  const int dH = kKnownStrideH > 0 ? kKnownStrideH : dH_;
+  const int dW = kKnownStrideW > 0 ? kKnownStrideW : dW_;
+
+  const int k_col = blockIdx.x % kW;
+  const int k_row = (blockIdx.x / kW) % kH;
+  const int k_frame = (blockIdx.x / kW / kH) % kT;
+  const int k_channel = blockIdx.x / kW / kH / kT;
+  scalar_t *result = &grad_kernel[k_channel][0][k_frame][k_row][k_col];
+
+  const int oT = grad_output.size(2);
+  const int oH = grad_output.size(3);
+  const int oW = grad_output.size(4);
+  const int iT = input.size(2);
+  const int iH = input.size(3);
+  const int iW = input.size(4);
+  const int channel_multiplier = grad_output.size(1) / input.size(1);
+  const int in_channel = k_channel / channel_multiplier;
+
+  extern __shared__ int sdata_raw[];
+  scalar_t* sdata = reinterpret_cast<scalar_t*>(sdata_raw);
+
+  if (k_channel >= kC) {
+    return;
+  }
+
+  const int laneid = threadIdx.x % C10_WARP_SIZE;
+  const int warpid = threadIdx.x / C10_WARP_SIZE;
+  const int nwarps = blockDim.x / C10_WARP_SIZE;
+
+  scalar_t grad = (scalar_t)0;
+  int batch = warpid / oT;
+  int gout_frame = warpid - batch * oT;
+  for (int outer_pos = warpid; outer_pos < input.size(0) * oT; 
+      outer_pos += nwarps, gout_frame += nwarps) {
+    while (gout_frame >= oT) { gout_frame -= oT; batch ++; }
+
+    const int in_frame = (gout_frame * dT) + (k_frame * dilationT) - pT;
+
+    if (in_frame < 0 || in_frame >= iT) {
+      continue;
+    }
+
+    const scalar_t* gout_ptr = grad_output[batch][k_channel][gout_frame].data() + laneid;
+    const scalar_t* input_ptr = input[batch][in_channel][in_frame].data();
+
+    int gout_row = laneid / oW;
+    int gout_col = laneid - gout_row * oW;
+
+    for (; gout_row < oH; ) {
+      const scalar_t op1 = __ldg(gout_ptr);
+      gout_ptr += C10_WARP_SIZE;
+
+      const int in_col = (gout_col * dW) + (k_col * dilationW) - pW;
+      const int in_row = (gout_row * dH) + (k_row * dilationH) - pH;
+      const int in_pos = in_row * iW + in_col;
+
+      scalar_t op2 = (scalar_t)0;
+      if (in_col >= 0 && in_col < iW && in_row >= 0 && in_row < iH) {
+        op2 = __ldg(input_ptr + in_pos);
+      }
+
+      gout_col += C10_WARP_SIZE;
+      while (gout_col >= oW) {
+        gout_col -= oW; gout_row ++;
+      }
+
+      grad += op1 * op2;
+    }
+  }
+  
+  sdata[threadIdx.x] = grad;
+  __syncthreads();
+ 
+  assert(__popc(blockDim.x) == 1);
```

Comment:
if you really need this assert use CUDA_KERNEL_ASSERT, however, it can be done on the hostside.